File: piwigo/Settings/Cells/SwitchTableViewCell.m
//
//  SwitchTableViewCell.m
//  piwigo
//
//  Created by Spencer Baker on 3/24/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//

#import "SwitchTableViewCell.h"

@interface SwitchTableViewCell()

@end

@implementation SwitchTableViewCell

-(instancetype)init
{
	self = [super init];
	if(self)
	{
		self.backgroundColor = [UIColor piwigoColorCellBackground];
		
		self.leftLabel = [UILabel new];
		self.leftLabel.translatesAutoresizingMaskIntoConstraints = NO;
		self.leftLabel.font = [UIFont piwigoFontNormal];
		self.leftLabel.textColor = [UIColor piwigoColorLeftLabel];
		self.leftLabel.adjustsFontSizeToFitWidth = NO;
		self.leftLabel.numberOfLines = 2;
		self.leftLabel.preferredMaxLayoutWidth = 150;
		[self.contentView addSubview:self.leftLabel];
		
		self.cellSwitch = [UISwitch new];
		self.cellSwitch.translatesAutoresizingMaskIntoConstraints = NO;
        self.cellSwitch.thumbTintColor = [UIColor piwigoColorThumb];
        self.cellSwitch.onTintColor = [UIColor piwigoColorOrange];
		[self.contentView addSubview:self.cellSwitch];
		[self.cellSwitch addTarget:self action:@selector(switchChanged) forControlEvents:UIControlEventValueChanged];
		
		[self setupConstraints];
	}
	return self;
}

-(void)setupConstraints
{
	NSDictionary *views = @{
							@"label" : self.leftLabel,
							@"switch" : self.cellSwitch
							};
	
	[self.contentView addConstraint:[NSLayoutConstraint constraintCenterHorizontalView:self.leftLabel]];
	[self.contentView addConstraint:[NSLayoutConstraint constraintCenterHorizontalView:self.cellSwitch]];
	
    if (@available(iOS 11, *)) {
        [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-[label]-[switch]-|"
																			 options:kNilOptions
																			 metrics:nil
																			   views:views]];
    } else {
        [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-15-[label]-[switch]-15-|"
                                                                                 options:kNilOptions
                                                                                 metrics:nil
                                                                                   views:views]];
    }
}

-(void)switchChanged
{
	if(self.cellSwitchBlock)
	{
		self.cellSwitchBlock(self.cellSwitch.isOn);
	}
}

@end

File: piwigo/Settings/Cells/SwitchTableViewCell.swift
//
//  SwitchTableViewCell.swift
//  piwigo
//
//  Created by Spencer Baker on 3/24/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//
//  Converted to Swift 5 by Eddy Lelièvre-Berna on 12/04/2020.
//

import UIKit

typealias CellSwitchBlock = (Bool) -> Void

class SwitchTableViewCell: UITableViewCell {

    @IBOutlet weak var switchName: UILabel!
    @IBOutlet weak var cellSwitch: UISwitch!
    @IBOutlet weak var topMargin: NSLayoutConstraint!
    @IBOutlet weak var bottomMargin: NSLayoutConstraint!

    var cellSwitchBlock: CellSwitchBlock?

    func configure(with title:String) {

        // Background color and aspect
        backgroundColor = PwgColor.cellBackground
        topMargin.constant = TableViewUtilities.vertMargin
        bottomMargin.constant = TableViewUtilities.vertMargin

        // Switch name
        switchName.textColor = PwgColor.leftLabel
        switchName.text = title

        // Switch appearance and action
        cellSwitch.thumbTintColor = PwgColor.thumb
        cellSwitch.onTintColor = PwgColor.orange
    }

    @IBAction func switchChanged(_ sender: Any) {
        (cellSwitchBlock ?? {_ in return})(cellSwitch.isOn)
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        switchName.text = ""
    }
}
